### DISCUSSION TITLE
Predict objects segmentation even when camera don't see, based on there objects on the scene

### DISCUSSION THREAD

**Discussion by damavand1** [ultralytics/ultralytics#9228] [General]
Hi again

Sometimes we don't see certain objects in the picture, but based on other objects, we can guess the invisible objects and their locations.
For example in lane line detection in this picture we don't see some parts of line, and humans guess invisible parts from some cues like other lane lines, building and trees

![11111](https://github.com/ultralytics/ultralytics/assets/104978921/e36cfdf7-6180-49d9-af1b-fa81929d4c64)

somethings like this:
[https://youtu.be/Ucp0TTmvqOE?t=7204](https://youtu.be/Ucp0TTmvqOE?t=7204)

as you see, Elon says: the network is predicting paths it can't even see
[https://youtu.be/Ucp0TTmvqOE?t=8076](https://youtu.be/Ucp0TTmvqOE?t=8076)

now, the question is how can train a network that predicts, when it can't even see ?

Thanks

**Comment by pderrenger** [comment]
Hi there,

Great question! To train a network to predict objects or paths it can't directly see, you'll need to focus on context-driven learning. In essence, the model needs to be trained on a dataset where it can learn the relationships between visible cues and the 'invisible' objects or paths. This requires a dataset rich in contextual information and possibly additional annotations for the invisible parts.

For instance, in lane detection, the dataset should include images where parts of the lane are obscured but annotated as if they were visible, based on the context provided by surrounding cues (like continuing lane lines, roadside patterns, etc.). The model can then learn to infer the presence of these obscured parts during training.

Here's a simplified example of how you might approach such a training scenario:

```python
from ultralytics import YOLO

# Initialize the model
model = YOLO('path/to/config.yaml')

# Train the model with custom dataset containing annotated 'invisible' paths
model.train(data='path/to/custom_dataset.yaml', epochs=100, imgsz=640)
```

In the `custom_dataset.yaml`, you'd include annotations for both the visible parts of the lanes and the guessed/estimated continuation of lanes that aren't directly visible. This way, the network learns to understand the context and can start predicting these 'invisible' objects or paths based on that understanding.

Remember, the key here is the quality and comprehensiveness of the dataset and its annotations. The network's ability to predict unseen objects heavily relies on the patterns and relationships it was trained on.

Hope this helps! Let us know if you have more questions. 🚀

**Reply by damavand1** [reply]
Hi pderrenger

I'm try to learn context-driven learning

Special thanks

**Reply by pderrenger** [reply]
Hello pderrenger,

That's fantastic to hear! Context-driven learning can be a powerful approach in enhancing your models, especially for tasks where understanding the scene context is crucial. 🌟

Here’s a quick example to get started with context-driven learning using Ultralytics YOLOv8:

```python
from ultralytics import YOLO

# Load your YOLO model
model = YOLO('yolov8n.pt')  # For instance

# Assuming you have your structured dataset ready
results = model.train(data='your_custom_dataset.yaml')
```

Remember, the richness of your dataset in terms of contextual relationships is key. Dive into the dataset, ensure it's well-annotated with the context you're targeting, and get the training going.

Happy learning, and do not hesitate to reach out if you need further assistance! 🚀
